fendermint/vm/message/src/ipc.rs

```diff
@@ -0,0 +1,169 @@
+// Copyright 2022-2023 Protocol Labs
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use cid::Cid;
+use fvm_shared::{address::Address, clock::ChainEpoch, crypto::signature::Signature};
+use ipc_sdk::subnet_id::SubnetID;
+use serde::{Deserialize, Serialize};
+
+/// Messages involved in InterPlanetary Consensus.
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
+#[allow(clippy::large_enum_variant)]
+pub enum IpcMessage {
+    /// A bottom-up checkpoint coming from a child subnet "for resolution", relayed by a user of the parent subnet for a reward.
+    ///
+    /// The reward can be given immediately upon the validation of the quorum certificate in the checkpoint,
+    /// or later during execution, once data availability has been confirmed.
+    BottomUpResolve(SignedRelayedMessage<CertifiedMessage<BottomUpCheckpoint>>),
+
+    /// A bottom-up checkpoint proposed "for execution" by the parent subnet validators, provided that the majority of them
+    /// have the data available to them, already resolved.
+    ///
+    /// To prove that the data is available, we can either use the ABCI++ "process proposal" mechanism,
+    /// or we can gossip votes using the _IPLD Resolver_ and attach them as a quorum certificate.
+    BottomUpExec(CertifiedMessage<BottomUpCheckpoint>),
+
+    // TODO
+    TopDown,
+}
+
+/// A message relayed by a user on the current subnet.
+///
+/// The relayer pays for the inclusion of the message in the ledger,
+/// but not necessarily for the execution of its contents.
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
+pub struct RelayedMessage<T> {
+    /// The relayed message.
+    pub message: T,
+    /// The address (public key) of the relayer in the current subnet.
+    pub relayer: Address,
+    /// The nonce of the relayer in the current subnet.
+    pub sequence: u64,
+}
+
+/// Relayed messages are signed by the relayer, so we can rightfully charge them message inclusion costs.
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
+pub struct SignedRelayedMessage<T> {
+    /// The relayed message with the relayer identity.
+    pub message: RelayedMessage<T>,
```

Comment:
Man, I hate that Filecoin uses `message` for `transaction` :facepalm:, it is so confusing sometimes. Here we are sending an arbitrary message, but then in the `bottom_up_messages` of the checkpoint we have the `Cid` of the bottom-up transactions being propagated, right?